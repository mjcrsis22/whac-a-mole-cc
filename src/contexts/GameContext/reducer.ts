import {
  GAME_LEVEL,
  GameSetting,
  GAME_SETTINGS_BY_LEVEL,
} from '../../constants/game';

export interface State {
  started: boolean;
  levelSetting: GameSetting;
  points: number;
}

export const INITIAL_STATE: State = {
  started: false,
  levelSetting: GAME_SETTINGS_BY_LEVEL[GAME_LEVEL.LOW],
  points: 0,
};

export enum ActionTypes {
  SET_STARTED = 'SET_STARTED',
  SET_LEVEL = 'SET_LEVEL',
  SET_POINTS = 'SET_POINTS',
  SET_DEFAULT = 'SET_DEFAULT',
}

export type SetStarted = {
  type: ActionTypes.SET_STARTED;
  payload: boolean;
};

export type SetLevel = {
  type: ActionTypes.SET_LEVEL;
  payload: GameSetting;
};

export type SetPoints = {
  type: ActionTypes.SET_POINTS;
  payload: number;
};

export type SetDefault = {
  type: ActionTypes.SET_DEFAULT;
};

export type Actions = SetStarted | SetLevel | SetPoints | SetDefault;

export function reducer(state: State, action: Actions) {
  switch (action.type) {
    case ActionTypes.SET_STARTED:
      return {
        ...state,
        started: action.payload,
      };
    case ActionTypes.SET_LEVEL:
      return {
        ...state,
        levelSetting: action.payload,
      };
    case ActionTypes.SET_POINTS:
      return {
        ...state,
        points: action.payload,
      };
    default:
      return state;
  }
}
