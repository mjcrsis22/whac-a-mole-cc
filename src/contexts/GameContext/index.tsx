import { createContext, Dispatch, ReactNode, useReducer } from 'react';
import { State, INITIAL_STATE, Actions, reducer } from './reducer';

export interface GameContextType {
  state: State;
  dispatch: Dispatch<Actions>;
}

// eslint-disable-next-line react-refresh/only-export-components
export const GameContext = createContext<GameContextType>({
  state: INITIAL_STATE,
  dispatch: () => {},
});

interface GameContextProviderProps {
  children: ReactNode;
}

export default function GameContextProvider({
  children,
}: GameContextProviderProps) {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);

  return (
    <GameContext.Provider value={{ state, dispatch }}>
      {children}
    </GameContext.Provider>
  );
}
