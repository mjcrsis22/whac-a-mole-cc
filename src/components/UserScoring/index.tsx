import { useContext } from 'react';
import { GameContext } from '../../contexts/GameContext';
import './styles.scss';

function UserScoring() {
  const {
    state: { points },
  } = useContext(GameContext);
  return <>{points}</>;
}

export default UserScoring;
