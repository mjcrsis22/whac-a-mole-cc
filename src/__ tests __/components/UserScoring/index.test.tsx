import { useContext } from 'react';
import { render } from '@testing-library/react';
import GameContextProvider from '../../../contexts/GameContext';
import { INITIAL_STATE } from '../../../contexts/GameContext/reducer';
import UserScoring from '../../../components/UserScoring';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

const mockedContext = {
  state: INITIAL_STATE,
  dispatch: jest.fn(),
};

const mockedUseContext = useContext as jest.MockedFn<typeof useContext>;

describe('UserScoring component', () => {
  beforeEach(() => {
    mockedUseContext.mockReturnValue(mockedContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    const { container } = render(
      <GameContextProvider>
        <UserScoring />
      </GameContextProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
