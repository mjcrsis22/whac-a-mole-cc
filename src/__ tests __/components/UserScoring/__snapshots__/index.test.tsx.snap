// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserScoring component Renders correctly 1`] = `
<div>
  0
</div>
`;
